### PR TITLE
feat: add cache invalidation for icons & avatars

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Backend aspect of #5795

### CODE REVIEW

File: rotkehlchen/api/rest.py [hunk 1]
```diff
@@ -4234,3 +4234,41 @@ def get_ens_avatar(self, ens_name: str, match_header: Optional[str]) -> Response
...
+
+                if entry.is_file():
+                    entry.unlink()
+
+            return api_response(OK_RESULT)
+
+        if cache_type == 'icons':
+            for identifier in entries:
+                self.rotkehlchen.icon_manager.delete_icon(Asset(identifier))
+            return api_response(OK_RESULT)
```
Comment (nebolax) [inline rotkehlchen/api/rest.py:4267]: Question. Shouldn't an error response be returned if no icon/avatar was found for the requested identifier?

File: rotkehlchen/api/rest.py [hunk 2]
```diff
@@ -4234,3 +4234,41 @@ def get_ens_avatar(self, ens_name: str, match_header: Optional[str]) -> Response
...
+
+        avatars_to_delete = [cache_dir / f'{i}.png' for i in entries]
```
Comment (nebolax) [inline rotkehlchen/api/rest.py:4269]:
```suggestion
        avatars_to_delete = [cache_dir / f'{ens_name}.png' for ens_name in entries]
```

File: rotkehlchen/api/v1/schemas.py [hunk 3]
```diff
@@ -98,6 +98,7 @@
 from rotkehlchen.utils.misc import create_order_by_rules_list, ts_now
 from rotkehlchen.utils.mixins.serializableenum import SerializableEnumMixin
 
+from ...errors.asset import UnknownAsset
```
Comment (nebolax) [inline rotkehlchen/api/v1/schemas.py:101]: I think this should better be an absolute import

File: rotkehlchen/api/v1/schemas.py [hunk 4]
```diff
@@ -2939,7 +2940,7 @@ def __init__(self, db: 'DBHandler') -> None:
         self.db = db
 
     @validates_schema
-    def validate_schema(
+    def validate_schema(  # pylint: disable=no-self-use
```
Comment (nebolax) [inline rotkehlchen/api/v1/schemas.py:2943]: Why this change?

File: rotkehlchen/icons.py [hunk 5]
```diff
@@ -91,6 +91,9 @@ def __init__(self, data_dir: Path, coingecko: Coingecko) -> None:
         self.custom_icons_dir.mkdir(parents=True, exist_ok=True)
         self.failed_asset_ids: set[str] = set()
 
+        avatars_dir = self.icons_dir / 'avatars'
```
Comment (nebolax) [inline rotkehlchen/icons.py:94]: Why is this needed? Existence of avatars directory is already ensured (at the moment of querying an avatar)

File: rotkehlchen/tests/api/test_misc.py [hunk 6]
```diff
@@ -409,3 +409,84 @@ def test_query_all_chain_ids(rotkehlchen_api_server):
     response = requests.get(api_url_for(rotkehlchen_api_server, 'allevmchainsresource'))
     result = assert_proper_response_with_result(response)
     assert result == [{'id': chain.value, 'name': chain.to_name()} for chain in ChainID]
+
+
+def test_cache_deletion(rotkehlchen_api_server):
```
Comment (nebolax) [inline rotkehlchen/tests/api/test_misc.py:414]:
```suggestion
def test_images_cache_deletion(rotkehlchen_api_server):
```

Comment (nebolax) [inline rotkehlchen/tests/api/test_misc.py:414]: Also a small docstring won't hurt

File: rotkehlchen/tests/api/test_misc.py [hunk 7]
```diff
@@ -409,3 +409,84 @@ def test_query_all_chain_ids(rotkehlchen_api_server):
...
+    icons_dir = rotkehlchen_api_server.rest_api.rotkehlchen.icon_manager.icons_dir
+    avatars_dir = icons_dir / 'avatars'
```
Comment (nebolax) [inline rotkehlchen/tests/api/test_misc.py:416]: Maybe move `avatars_dir` to `icon_manager`? So that it can be accessed as a property too.

File: rotkehlchen/api/rest.py (hunk 1 above)
Comment (prettyirrelevant) [reply rotkehlchen/api/rest.py:4267]: I don't think that should be the case since the frontend doesn't know if an avatar/icon exists in the first instance

File: rotkehlchen/api/v1/schemas.py (hunk 3 above)
Comment (prettyirrelevant) [reply rotkehlchen/api/v1/schemas.py:101]: yeah, a mistake from my ide during an auto-import

File: rotkehlchen/tests/api/test_misc.py (hunk 7 above)
Comment (prettyirrelevant) [reply rotkehlchen/tests/api/test_misc.py:416]: hmm okay.

File: rotkehlchen/api/rest.py [hunk 8]
```diff
@@ -4234,3 +4234,41 @@ def get_ens_avatar(self, ens_name: str, match_header: Optional[str]) -> Response
...
+        """Removes cached items from the filesystem."""
+        if cache_type == 'icons':
+            cache_dir = self.rotkehlchen.icon_manager.icons_dir
+        else:  # can only be avatars
+            cache_dir = self.rotkehlchen.icon_manager.icons_dir / 'avatars'
+
+        # delete every file in the cache directory.
+        if entries is None or len(entries) == 0:
+            for entry in cache_dir.iterdir():
+                # don't delete avatars dir when clearing icons cache
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/api/rest.py:4252]: Yeah, this is cool. Was surprised avatars were placed in icons directory. So a data migration.

File: rotkehlchen/tests/api/test_misc.py (hunk 7 above)
Comment (prettyirrelevant) [reply rotkehlchen/tests/api/test_misc.py:416]: But that has nothing to do with icons, so I'm not so sure about that.

File: rotkehlchen/api/rest.py (hunk 1 above)
Comment (nebolax) [reply rotkehlchen/api/rest.py:4260]: Yes, probably makes sense. But pinging @rotki/frontend just to make sure.

File: rotkehlchen/tests/api/test_misc.py (hunk 7 above)
Comment (nebolax) [reply rotkehlchen/tests/api/test_misc.py]: Yes, has nothing to do with icons indeed, however our avatars logic is already tied to icons logic. So I would ask @LefterisJP about this

File: docs/api.rst [hunk 9]
```diff
@@ -12408,3 +12408,36 @@ Get avatar for an ens name
    :statuscode 400: Invalid ens name provided (it doesn't end with `.eth`).
    :statuscode 404: There is no avatar set for the given ens name.
    :statuscode 409: Avatar was requested for an ens name that is not currently in the database or we couldn't query the blockchain.
+
+
+Clear Icons/Avatars Cache
+============================
+
+.. http:post:: /api/(version)/cache/images/<cache_type>/clear
```
Comment (LefterisJP) [inline docs/api.rst:12416]: If you are adding a new `/cache` endpoint perhaps it would make sense to move other cache/data cleaning there?

File: docs/api.rst [hunk 10]
```diff
@@ -12408,3 +12408,36 @@ Get avatar for an ens name
...
+   **Example Request**
+
+    .. http:example:: curl wget httpie python-requests
+
+      POST /api/1/cache/icons/clear HTTP/1.1
+      Host: localhost:5042
+      Content-Type: application/json;charset=UTF-8
+
+      {
+         "entries": ["ETH", "BTC"]
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply docs/api.rst:12429]: no it's possible and you can keep the validation. You just make the sole argument be a list. But this is also fine. No need to change.

File: docs/api.rst [hunk 11]
```diff
@@ -12408,3 +12408,36 @@ Get avatar for an ens name
...
+   **Example Response**:
+
+   .. sourcecode:: http
+
+      HTTP/1.1 200 OK
+      Content-Type: application/json
+
+   :statuscode 200: The cache items were cleared successfully.
+   :statuscode 400: Entries provided do not match the ``cache_type`` specified. Invalid ``cache_type`` provided.
+    :statuscode 500: Internal rotki error
```
Comment (LefterisJP) [inline docs/api.rst:12443]: At least in github preview this seems like it has an extra space?

```suggestion
   :statuscode 500: Internal rotki error
```

File: rotkehlchen/api/rest.py [hunk 12]
```diff
@@ -4196,7 +4196,7 @@ def get_ens_avatar(self, ens_name: str, match_header: Optional[str]) -> Response
 
         Also supports etag mechanism that helps with caching on the client side.
         """
-        avatars_dir = self.rotkehlchen.data_dir / 'icons' / 'avatars'
+        avatars_dir = self.rotkehlchen.data_dir / 'images' / 'avatars'
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:4199]: What are you doing? Why are you renaming this? Undo. The `icons` directory already exists in all the user's systems. You are not going to create another one. If we ever wanted to do changes there we would need a migration.

File: rotkehlchen/api/rest.py (hunk 1 above)
Comment (LefterisJP) [reply rotkehlchen/api/rest.py:4260]: For these stuff ask in discord. Seeing a ping here may take time. I also think it's fine if there is nothing special returned in case missing. Result is the same. No file for this identifier.

File: rotkehlchen/api/server.py [hunk 13]
```diff
@@ -314,6 +315,7 @@
         'per_timestamp_db_snapshots_resource',
     ),
     ('/notes', UserNotesResource),
+    ('/cache/<string:cache_type>/clear', ClearCacheResource),
```
Comment (LefterisJP) [inline rotkehlchen/api/server.py:318]: This looks different to the endpoint you described in api.rst. There you wrote:  

```
http:post:: /api/(version)/cache/images/<cache_type>/clear
```

```suggestion
    ('/cache/images/<string:cache_type>/clear', ClearImagesCacheResource),
```

So which one is it? Do you need the images or not? I would guess not, since we probably can keep one such endpoint for all types of caches?

File: rotkehlchen/data_migrations/migrations/migration_10.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import logging
+import shutil
+from typing import TYPE_CHECKING
+
+from rotkehlchen.logging import RotkehlchenLogsAdapter
```
Comment (LefterisJP) [inline rotkehlchen/data_migrations/migrations/migration_10.py:5]: Is this needed? Why can't you have `avatars` under `icons` for now?

I would prefer to do these kind of changes all together and also solve https://github.com/rotki/rotki/issues/4841 while we are at it.

File: rotkehlchen/tests/api/test_misc.py [hunk 15]
```diff
@@ -409,3 +409,85 @@ def test_query_all_chain_ids(rotkehlchen_api_server):
     response = requests.get(api_url_for(rotkehlchen_api_server, 'allevmchainsresource'))
     result = assert_proper_response_with_result(response)
     assert result == [{'id': chain.value, 'name': chain.to_name()} for chain in ChainID]
+
+
+def test_images_cache_deletion(rotkehlchen_api_server):
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_misc.py:414]: Set a temporary directory for this test since you are going to be creating and deleting files in that directory. 
```suggestion
@pytest.mark.parametrize('use_clean_caching_directory', [True])
def test_images_cache_deletion(rotkehlchen_api_server):
```

File: rotkehlchen/tests/api/test_misc.py [hunk 16]
```diff
@@ -409,3 +409,85 @@ def test_query_all_chain_ids(rotkehlchen_api_server):
...
+            'clearcacheresource',
+            cache_type='icons',
+        ),
+    )
+    assert_proper_response(response)
+    assert len([i for i in icons_dir.iterdir() if i.is_file()]) == 0
+    assert len([i for i in avatars_dir.iterdir() if i.is_file()]) == 1
+
+    # test avatars cache
+    # populate avatars dir
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_misc.py:461]:
```suggestion
    # populate avatars dir to test the cache deletion
```

File: rotkehlchen/tests/api/test_misc.py [hunk 17]
```diff
@@ -409,3 +409,85 @@ def test_query_all_chain_ids(rotkehlchen_api_server):
...
+    with open(f'{icons_dir}/ETH_small.png', 'wb') as f:
+        f.write(b'')
+
+    with open(f'{icons_dir}/BTC_small.png', 'wb') as f:
+        f.write(b'')
+
+    with open(f'{icons_dir}/AVAX_small.png', 'wb') as f:
+        f.write(b'')
+
+    # make sure this is not deleted.
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_misc.py:430]:
```suggestion
    # also add an avatar, to make sure it's not deleted when icons are
```

**Review by LefterisJP — COMMENTED** [review]
1. Many many tests are failing. Please don't send me PRs when this is the case.
2. Left some comments. Important: I don't think we need the new images directory yet. There is more stuff (linked an issue) we need to change at some point.

File: docs/api.rst [hunk 18]
```diff
@@ -12408,3 +12372,40 @@ Get avatar for an ens name
...
+
+      POST /api/1/cache/icons/clear HTTP/1.1
+      Host: localhost:5042
+      Content-Type: application/json;charset=UTF-8
+
+      {
+         "icons": ["ETH", "BTC"]
+      }
+
+   :reqjsonarr optional[string] icons: An array of the ass to be cleared from the cache. All items are deleted in the cache if ``null``. Only works with ``cache_type=icons``.
```
Comment (LefterisJP) [inline docs/api.rst:12396]: ass?

```suggestion
   :reqjsonarr optional[string] icons: An array of the ??? to be cleared from the cache. All items are deleted in the cache if ``null``. Only works with ``cache_type=icons``.
```

File: docs/api.rst [hunk 19]
```diff
@@ -2101,42 +2101,6 @@ Request creation of oracle price cache
    :statuscode 500: Internal rotki error
    :statuscode 502: The oracle could not be queried due to an error on their side.
 
-Delete an oracle price cache
```
Comment (LefterisJP) [inline docs/api.rst:2104]: This makes no sense. I did not mean to do it now. I meant to do it perhaps at some other point. This endpoint already exists and is used with other verbs. This change you are doing does not make sense then as is. Please ask before goign and doing stuff like that.

File: rotkehlchen/api/v1/resources.py [hunk 20]
```diff
@@ -2858,3 +2847,30 @@ class EnsAvatarsResource(BaseMethodView):
     @use_kwargs(get_schema, location='view_args')
     def get(self, ens_name: str) -> Response:
         return self.rest_api.get_ens_avatar(ens_name=ens_name, match_header=get_match_header())
+
+
+class ClearCacheResource(BaseMethodView):
+    post_schema = ClearCacheSchema()
+    icons_schema = ClearIconsCacheSchema()
+    avatars_schema = ClearAvatarsCacheSchema()
+    price_schema = NamedOracleCacheSchema()
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/resources.py:2856]: Undo this entire change please. It's not a good place while keeping the previous endpoint for price oracle caches for everything else

**Review by LefterisJP — COMMENTED** [review]
No no no. What you did by putting that price oracle cache in the new endpoint makes no sense.

**PLEASE** ask in discord either your colleagues or me when you try to do something like that.

You are wasting your time and mine by doing work that is not needed.

Why is it not a good idea?

1. You do not remove any endpoint. The endpoint for price oracle cache still exists and is accessed by other verbs. So you are force-moving a certain sub functionality only.
2. The schema verification becomes an ugly code salad. Instead of having the verification per endpoint nicely and clean you start some dynamic magic.

File: rotkehlchen/api/rest.py (hunk 1 above)
Comment (kelsos) [reply rotkehlchen/api/rest.py:4262]: We don't need any error code if the icon was missing in the first place. (Anything other than 200).

The only case IMO that would make sense to get an error is if the deletion fails, just to notify the user that the action was not successful.

File: rotkehlchen/api/v1/schemas.py [hunk 21]
```diff
@@ -2997,3 +2998,32 @@ def make_binance_savings_query(
 
 class EnsAvatarsSchema(Schema):
     ens_name = fields.String(required=True, validate=lambda x: x.endswith('.eth'))
+
+
+class ClearCacheSchema(Schema):
+    cache_type = fields.String(
+        required=True,
+        validate=[validate.OneOf({'icons', 'avatars', 'prices'})],
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/schemas.py:3006]:
```suggestion
        validate=[validate.OneOf({'icons', 'avatars')],
```
